File: realtube-python/app/routers/health.py
"""Health check endpoints: liveness and readiness probes."""

import time

import asyncpg
from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse

router = APIRouter(tags=["health"])

_start_time: float = time.time()


@router.get("/health/live")
async def live():
    return {"status": "ok"}


@router.get("/health/ready")
async def ready(request: Request):
    checks: dict = {}
    overall_status = "healthy"

    # Database check
    pool: asyncpg.Pool | None = getattr(request.app.state, "db_pool", None)
    if pool is not None:
        checks["database"] = await _check_db(pool)
    else:
        checks["database"] = {"status": "down", "error": "no pool"}

    if checks["database"]["status"] != "up":
        overall_status = "degraded"

    # Redis check
    cache_svc = getattr(request.app.state, "cache_service", None)
    rdb = cache_svc.client if cache_svc else None
    if rdb is not None:
        checks["redis"] = await _check_redis(rdb)
    else:
        checks["redis"] = {"status": "disabled"}

    if checks["redis"].get("status") not in ("up", "disabled") and overall_status == "healthy":
        overall_status = "degraded"

    uptime_seconds = int(time.time() - _start_time)

    resp = {
        "status": overall_status,
        "checks": checks,
        "uptime_seconds": uptime_seconds,
        "version": "1.0.0",
    }

    status_code = 200 if overall_status == "healthy" else 503
    return JSONResponse(content=resp, status_code=status_code)


async def _check_db(pool: asyncpg.Pool) -> dict:
    start = time.time()
    try:
        async with pool.acquire() as conn:
            await conn.fetchval("SELECT 1")
        latency_ms = int((time.time() - start) * 1000)
        return {"status": "up", "latency_ms": latency_ms}
    except Exception as e:
        latency_ms = int((time.time() - start) * 1000)
        return {"status": "down", "latency_ms": latency_ms, "error": str(e)}


async def _check_redis(rdb) -> dict:
    start = time.time()
    try:
        await rdb.ping()
        latency_ms = int((time.time() - start) * 1000)
        return {"status": "up", "latency_ms": latency_ms}
    except Exception as e:
        latency_ms = int((time.time() - start) * 1000)
        return {"status": "down", "latency_ms": latency_ms, "error": str(e)}

File: realtube-go/internal/handler/health.go
package handler

import (
	"context"
	"time"

	"github.com/gofiber/fiber/v3"
	"github.com/jackc/pgx/v5/pgxpool"
	"github.com/redis/go-redis/v9"
)

type HealthHandler struct {
	pool    *pgxpool.Pool
	rdb     *redis.Client
	startAt time.Time
}

func NewHealthHandler(pool *pgxpool.Pool, rdb *redis.Client) *HealthHandler {
	return &HealthHandler{
		pool:    pool,
		rdb:     rdb,
		startAt: time.Now(),
	}
}

// Live handles GET /health/live — liveness probe.
func (h *HealthHandler) Live(c fiber.Ctx) error {
	return c.JSON(fiber.Map{"status": "ok"})
}

// Ready handles GET /health/ready — readiness probe with dependency checks.
func (h *HealthHandler) Ready(c fiber.Ctx) error {
	ctx, cancel := context.WithTimeout(c.Context(), 3*time.Second)
	defer cancel()

	checks := make(fiber.Map)
	overallStatus := "healthy"

	// Database check
	checks["database"] = checkDB(ctx, h.pool)
	if dbCheck, ok := checks["database"].(fiber.Map); ok {
		if dbCheck["status"] != "up" {
			overallStatus = "degraded"
		}
	}

	// Redis check
	checks["redis"] = checkRedis(ctx, h.rdb)
	if redisCheck, ok := checks["redis"].(fiber.Map); ok {
		if redisCheck["status"] != "up" && overallStatus == "healthy" {
			overallStatus = "degraded"
		}
	}

	uptimeSeconds := int(time.Since(h.startAt).Seconds())

	resp := fiber.Map{
		"status":         overallStatus,
		"checks":         checks,
		"uptime_seconds": uptimeSeconds,
		"version":        "1.0.0",
	}

	status := fiber.StatusOK
	if overallStatus != "healthy" {
		status = fiber.StatusServiceUnavailable
	}

	return c.Status(status).JSON(resp)
}

func checkDB(ctx context.Context, pool *pgxpool.Pool) fiber.Map {
	start := time.Now()
	err := pool.Ping(ctx)
	latency := time.Since(start).Milliseconds()

	if err != nil {
		return fiber.Map{
			"status":     "down",
			"latency_ms": latency,
			"error":      "connection failed",
		}
	}
	return fiber.Map{
		"status":     "up",
		"latency_ms": latency,
	}
}

func checkRedis(ctx context.Context, rdb *redis.Client) fiber.Map {
	if rdb == nil {
		return fiber.Map{
			"status": "disabled",
		}
	}

	start := time.Now()
	err := rdb.Ping(ctx).Err()
	latency := time.Since(start).Milliseconds()

	if err != nil {
		return fiber.Map{
			"status":     "down",
			"latency_ms": latency,
			"error":      "connection failed",
		}
	}
	return fiber.Map{
		"status":     "up",
		"latency_ms": latency,
	}
}
